src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs

```diff
@@ -352,9 +352,9 @@ public override async Task ExecuteAsync(Func<Task> operation, CancellationToken
                     }
                 }
 
-                await this.ExecuteCoreAsync(async delegate
+                try
```

Comment:
I would be inclined to begin the `try` block immediately after declaring `releaser`.  It would be a shame if in the future some potentially-throwing code snuck in at line 354 and we ended up chasing this horrible bug all over again.

The same suggestion applies in all cases, of course.